src/lavinmqperf/mqtt/throughput.cr

```diff
@@ -0,0 +1,345 @@
+          @qos = v.to_i
+        end
+        @parser.on("-t topic", "--topic=name", "Topic name (default perf-test)") do |v|
+          @topic = v
+        end
+        @parser.on("-r pub-rate", "--rate=number", "Max publish rate (default 0)") do |v|
+          @rate = v.to_i
+        end
+        @parser.on("-R consumer-rate", "--consumer-rate=number", "Max consume rate (default 0)") do |v|
+          @consume_rate = v.to_i
+        end
+        @parser.on("-j", "--json", "Output result as JSON") do
+          @json_output = true
+        end
+        @parser.on("-z seconds", "--time=seconds", "Only run for X seconds") do |v|
+          @timeout = Time::Span.new(seconds: v.to_i)
+        end
+        @parser.on("-q", "--quiet", "Quiet, only print the summary") do
+          @quiet = true
+        end
+        @parser.on("-C messages", "--pmessages=messages", "Publish max X number of messages") do |v|
+          @pmessages = v.to_i
+        end
+        @parser.on("-D messages", "--cmessages=messages", "Consume max X number of messages") do |v|
+          @cmessages = v.to_i
+        end
+        @parser.on("--random-bodies", "Each message body is random") do
+          @random_bodies = true
+        end
+        @parser.on("--retain", "Set retain flag on published messages") do
+          @retain = true
+        end
+        @parser.on("--clean-session", "Use clean session") do
+          @clean_session = true
+        end
+        @parser.on("-u uri", "--uri=uri", "MQTT broker URI (default mqtt://localhost:1883)") do |v|
+          @uri = URI.parse(v)
+        end
+      end
+
+      @pubs = 0_u64
+      @consumes = 0_u64
+      @stopped = false
+      @counter_lock = Mutex.new
+
+      private def create_client(id : Int32, role : String) : {TCPSocket, LavinMQ::MQTT::IO}
+        host = @uri.host || "localhost"
+        port = @uri.port || 1883
+        user = @uri.user || "guest"
+        password = @uri.password || "guest"
+
+        socket = TCPSocket.new(host, port)
+        socket.keepalive = true
+        socket.tcp_nodelay = false
+        socket.sync = true
+        io = LavinMQ::MQTT::IO.new(socket)
+
+        client_id = "#{role}-#{id}"
+        connect_packet = LavinMQ::MQTT::Connect.new(
+          client_id: client_id,
+          clean_session: @clean_session,
+          keepalive: 30_u16,
+          username: user,
+          password: password.to_slice,
+          will: nil
+        )
+
+        connect_packet.to_io(io)
+
+        response = LavinMQ::MQTT::Packet.from_io(io)
+        unless response.is_a?(LavinMQ::MQTT::Connack) &&
+               response.as(LavinMQ::MQTT::Connack).return_code == LavinMQ::MQTT::Connack::ReturnCode::Accepted
+          socket.close rescue nil
+          raise "Failed to connect: #{response.inspect}"
+        end
+
+        {socket, io}
+      rescue ex : Exception
+        socket.try &.close rescue nil
+        puts "Error creating connection: #{ex.message}"
+        raise ex
```

Comment:
We log and raise here, catch it in `consume` or `pub`  which logs and raises which in the end will crash the fiber with a stacktrace. 
Should be maybe just not rescue it and in the spawn block rescue any error and print the `e.message` 

What I'm thinking is, is there an idea to how to handle exceptions or have we just added rescue when needed? :)

Reply:
Haha yes, no plan just vibecoding ^^ you're right! i'll clean this up

Reply:
check out commit: f0cd62819a1523953f2c6fe29e2926294499dcf2
is this what you mean? :)

Reply:
Or rather we need to do it like in amqp, where reconnect_on_disconnect handles everything: 68a56c0f17d38cab51582dec8b3c680798d65775

Reply:
Yes, Both the commits are great improvments and having it in `reconnect_on_disconnect` makes sense. You just have to remember to set the message on the exception but that is no problem.